### PR TITLE
Be able to execute a script on host and node source

### PR DESCRIPTION
**Pull request by jrochas**

### CODE REVIEW

File: rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/RMController.java [hunk 1]
```diff
@@ -1296,24 +1294,54 @@ public void onUncaughtException(Throwable e) {
...
+                scriptResultStr.append(parseScriptResult(currentScriptResultJsonObject)).append("<br/>");
+            }
+            return scriptResultStr.toString();
+        } else {
+            return parseScriptResult(scriptResultJsonObject);
+        }
+    }
+
+    private String parseScriptResult(JSONObject scriptResult) {
         StringBuilder scriptResultStr = new StringBuilder();
```
Comment (marcocast) [inline rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/RMController.java]: a bit too complex to read....is it possible to refactor it a bit to increase readability?

File: rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/RMController.java [hunk 2]
```diff
@@ -1337,7 +1365,59 @@ public void onFailure(Throwable caught) {
...
+                                   engine,
+                                   nodeSourceName,
+                                   new AsyncCallback<String>() {
+                                       public void onFailure(Throwable caught) {
+                                           String msg = JSONUtils.getJsonErrorMessage(caught);
+                                           LogModel.getInstance()
+                                                   .logImportantMessage("Failed to execute a script " + script +
+                                                                        " on node source " + nodeSourceName + " : " +
+                                                                        msg);
+                                           if (msg.equals("HTTP 500 Internal Server Error")) {
```
Comment (marcocast) [inline rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/RMController.java]: can you create a separate private method and delegate to it this error management? like handleError(String msg).....

File: rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/RMController.java [hunk 3]
```diff
@@ -1337,7 +1365,59 @@ public void onFailure(Throwable caught) {
...
+        rm.executeHostScript(LoginModel.getInstance().getSessionId(),
+                             script,
+                             engine,
+                             host,
+                             new AsyncCallback<String>() {
+                                 public void onFailure(Throwable caught) {
+                                     String msg = JSONUtils.getJsonErrorMessage(caught);
+                                     LogModel.getInstance().logImportantMessage("Failed to execute a script " + script +
+                                                                                " on host " + host + " : " + msg);
+                                     if (msg.equals("HTTP 500 Internal Server Error")) {
```
Comment (marcocast) [inline rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/RMController.java]: you can then reuse the handleError method as before

File: rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/RMController.java (hunk 1 above)
Comment (jrochas) [reply rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/RMController.java]: Yes, I created a method for each main entity parsed.

File: rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/RMController.java (hunk 2 above)
Comment (jrochas) [reply rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/RMController.java]: Yes, done.

File: rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/RMController.java (hunk 3 above)
Comment (jrochas) [reply rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/RMController.java]: Yes.

File: rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/NodeSource.java [hunk 4]
```diff
@@ -254,6 +254,14 @@ public static String generateId(String sourceName, String hostName) {
...
         }
 
+        public String getIcon() {
+            if (this.isVirtual()) {
+                return RMImages.instance.host_virtual_16().getSafeUri().asString();
+            } else {
+                return RMImages.instance.host_16().getSafeUri().asString();
+            }
+        }
+
```
Comment (nebilbenmabrouk) [inline rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/NodeSource.java]: We can change it like this (only one line) :
return this.isVirtual ? RMImages.instance.host_virtual_16().getSafeUri().asString()  : RMImages.instance.host_16().getSafeUri().asString()

File: rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/RMController.java [hunk 5]
```diff
@@ -1296,25 +1294,84 @@ public void onUncaughtException(Throwable e) {
...
-        JSONObject scriptResult = this.parseJSON(json).isObject();
+    private String parseAllScriptResults(String jsonString) {
+        JSONValue allScriptResultsJson = this.parseJSON(jsonString);
+        JSONObject scriptResultJsonObject = allScriptResultsJson.isObject();
+        if (scriptResultJsonObject != null) {
+            return parseScriptResultAsJsonObject(scriptResultJsonObject);
+        } else {
+            return parseAllScriptResultsAsJsonArray(allScriptResultsJson);
+        }
+    }
```
Comment (nebilbenmabrouk) [inline rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/RMController.java]: Same comment as before (we can change the if statement to only one line)

File: rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/RMController.java [hunk 6]
```diff
@@ -1296,25 +1294,84 @@ public void onUncaughtException(Throwable e) {
...
-        JSONObject exception = scriptResult.get("exception").isObject();
-        while (exception != null && exception.get("cause") != null && exception.get("cause").isObject() != null) {
-            exception = exception.get("cause").isObject();
+    private String parseAllScriptResultsAsJsonArray(JSONValue allScriptResultsJson) {
+        StringBuilder allScriptResultsOutput = new StringBuilder();
+        JSONArray scriptResultJsonArray = allScriptResultsJson.isArray();
+
+        JSONValue scriptResultJson;
+        JSONObject scriptResultJsonObject;
+        for (int i = 0; i < scriptResultJsonArray.size(); i++) {
```
Comment (nebilbenmabrouk) [inline rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/RMController.java]: We can change as below (but I'm not sure it works)
for ( JSONValue scriptResultJson : scriptResultJsonArray) {

File: rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/NodeSource.java (hunk 4 above)
Comment (jrochas) [reply rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/NodeSource.java]: Applied.

File: rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/RMController.java (hunk 6 above)
Comment (jrochas) [reply rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/RMController.java]: No it doesn't, unfortunately it is not iterable :-)

File: rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/RMController.java [hunk 7]
```diff
@@ -1325,21 +1382,59 @@ public void executeScript(final String script, final String engine, final String
...
                                  }
                              });
     }
 
+    private void reportScriptExecutionFailure(Throwable caught, String script, String scriptTarget,
+            Callback<String, String> syncCallBack) {
+        String msg = JSONUtils.getJsonErrorMessage(caught);
+        LogModel.getInstance()
+                .logImportantMessage("Failed to execute a script " + script + " on " + scriptTarget + ": " + msg);
+        if (msg.equals("HTTP 500 Internal Server Error")) {
```
[the remark this answers is not in the document]
Comment (jrochas) [reply rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/RMController.java]: Applied.

**Pull request merged**